Es netssl

jocelyn opened the pull request:
for review

jvelilla committed:
Initial import EVP GCM mode.

jvelilla committed:
Updated SSL_EVP interface : added encryption features.
Added example showing encrypt- decrypt using aes gcm 256 based on OpenSSL
example.

jvelilla committed:
Updated Decrypt example using byte array converter.

jvelilla committed:
Initial SSL_GCM high level API.
Added test case gcm decrypt

jocelyn committed:
Made BYTE_ARRAY_CONVERTER as hidden (implementation) class.
Merge openssl.ecf and test into openssl.ecf.
Splitted the APPLICATION.make feature.
Removed a few warnings.

jvelilla committed:
Merge pull request #1 from jocelyn/es_openssl_crypto

Made BYTE_ARRAY_CONVERTER as hidden (implementation) class.

jvelilla committed:
Updated OpenSSL crypto cluster, added a new cluster `wrapper' to have low level
code wrapping C code.
Initial commit gcm encrypt (work in progress).

jvelilla committed:
Update OpenSSL libcrypto interface.

jvelilla committed:
Updated SSL crypt algorithms.
Added new interfaces to represent the different modes and algorithm supported
by OpenSSL and a generic way to encrypt/decrypt.

jvelilla committed:
Updated OpenSSL crypto targeting Symmetric encryption (ie AES with GCM).
Added ERROR abstractions to handle OpenSSL errors.
Updated low level api to handle errors at the moment using Exceptions.
Added test cases to show AES with GCM mode encrypt/decrypt.

jvelilla committed:
Renamed cluster experimental as to_remove.

jvelilla committed:
Updated SSL_CIPHER_CONTEXT_EXTERNAL implementation with the
SSL_CIPHER_CONTEXT interface.

jvelilla committed:
Merge branch 'master' of https://github.com/EiffelSoftware/EiffelStudio into es_openssl_crypto

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

jvelilla committed:
Updated OpenSSL library
SSL class
Added `ssl_get_error` to get the result code of for a preceding call to SSL_connect(), SSL_accept(), SSL_do_handshake(), SSL_read(), SSL_peek(), or SSL_write() on ssl.
Added externals features to get the value of different error codes:
SSL_ERROR_WANT_READ, SSL_ERROR_NONE, SSL_ERROR_ZERO_RETURN, SSL_ERROR_WANT_WRITE
Added external feature `ssl_pending` to return the number of bytes with are available
inside SSL for inmediate read.

Updated NetSSL library to support non-blocking reads and writes.
Added a non blocking client/server example.

jocelyn commented on Src/unstable/library/network/openssl/openssl.ecf:
```diff
@@ -1,7 +1,7 @@
 	<target name="openssl">
-			<description>Eiffel OpenSSL Wrap library.
+		<description>Eiffel OpenSSL Wrap library.
       Copyright (c) 1984-2006, Eiffel Software and others.
```
copyright year should be updated

jvelilla replied:
done.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/algorithms/ssl_algorithm.e:
```diff
@@ -0,0 +1,45 @@
+	verify_key_size: BOOLEAN
+			-- has the current algorithm a valid key size?
+		do
+			if key_sizes.has (key_size) then
```
Result := key_sizes.has (key_size)

jvelilla replied:
done

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/interface/implementation/ssl_aead_cipher_context_impl.e:
```diff
@@ -0,0 +1,136 @@
+			Result := ctx.update (a_data)
+		end
+
+   update_into(a_data, a_buf: MANAGED_POINTER): INTEGER
```
bad indentation, and missing space before `(`

jvelilla replied:
done.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/interface/implementation/ssl_aead_cipher_context_impl.e:
```diff
@@ -0,0 +1,136 @@
+			else
+				updated := True
+				bytes_processed := bytes_processed + a_data_size
+--				if bytes_processed > ctx.mode.MAX_ENCRYPTED_BYTES then
```
either add a comment related to the commented lines, or remove those lines

jvelilla replied:
done

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/interface/implementation/ssl_aead_cipher_context_impl.e:
```diff
@@ -0,0 +1,136 @@
+
+feature -- Access
+
+	ctx: SSL_CIPHER_CONTEXT_EXTERNALS
```
should `ctx`  be protected  (export NONE) and renamed `implementation` or `externals` ?

jvelilla replied:
yes, I will export it to `{NONE}`,  but I like the name `ctx` or  `context` since it's more natural in this domain.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/interface/implementation/ssl_cipher_context_impl.e:
```diff
@@ -0,0 +1,70 @@
+
+   finalize: MANAGED_POINTER
+			-- <Precursor>
+		do
```
why not having a is_not_finalized precondition?

jvelilla replied:
Yes, added it.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/interface/ssl_aead_decryption_context.e:
```diff
@@ -0,0 +1,26 @@
+	EIS: "name=AEAD_DECRYPTION_CONTEXT", "src=https://cryptography.io/en/latest/hazmat/primitives/symmetric-encryption/?highlight=aeadciphercontext#cryptography.hazmat.primitives.ciphers.AEADDecryptionContext", "protocol=uri"
+
+deferred class
+	SSL_AEAD_DECRYPTION_CONTEXT
```
nowhere in the code I saw the description of AEAD ?

Authenticated encryption (AE) and authenticated encryption with associated data (AEAD, variant of AE)

jvelilla replied:
good point.  Done.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/interface/ssl_aead_encryption_context.e:
```diff
@@ -0,0 +1,19 @@
+feature -- Tag
+
+	tag: detachable MANAGED_POINTER
+			-- Returns tag value as bytes. This is only available after encryption is finalized.
```
why not having a precondition to express "is only available after encryption is finalized."

jvelilla replied:
done.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/interface/ssl_cipher_context.e:
```diff
@@ -0,0 +1,45 @@
+
+deferred class
+	SSL_CIPHER_CONTEXT
+
```
I am surprised to see the descendant of this class

SSL_CIPHER_CONTEXT
	SSL_AEAD_CIPHER_CONTEXT_IMPL
		SSL_AEAD_ENCRYPTION_CONTEXT_IMPL
	SSL_CIPHER_CONTEXT_EXTERNALS
	SSL_CIPHER_CONTEXT_IMPL

both _IMPL and _EXTERNALS ...
I haven't checked, but is it expected?

jvelilla replied:
CIPHER_CONTEXT is the high-level interface.  
_IMPL are implementations of other CONTEXT that add some aspects to the basic CIPHER_CONTEXT.
_EXTERNALS is the low-level implementation that also follows the high-level interface.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/modes/ssl_gcm_mode.e:
```diff
@@ -0,0 +1,71 @@
+
+feature -- Status Report
+
+	check_aes_key_length (a_algo: SSL_ALGORITHM): BOOLEAN
```
update to Eiffel style .. a verb should not return any value.
why not having `is_valid_aes_key (....): BOOLEAN` ?
or similar?

jvelilla replied:
that's ok.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/modes/ssl_gcm_mode.e:
```diff
@@ -0,0 +1,71 @@
+				l_algo.key_size > 256 and then l_algo.name.same_string ("AES")
+			then
+				Result := False
+					-- Only 128, 192, and 256 bit keys are allowed for this AES mode
```
What is l_algo.name is not "AES" .... should it really return True?

jvelilla replied:
good point, I will need to check the Python code.

jvelilla replied:
fixed.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/modes/ssl_mode.e:
```diff
@@ -0,0 +1,22 @@
+note
+	description: "Object representing modes that can be used for OpenSSL crypto algorithms"
```
modes or mode?
I guess an instance of SSL_MODE represents only one mode , no?
Also what is a "mode" ?

jvelilla replied:
it represents a mode like  `GCM`, `CCM`, `CTR`, an algorithm can be used in different modes. 
For example:
const EVP_CIPHER *EVP_aes_128_ctr(void);
const EVP_CIPHER *EVP_aes_128_ccm(void);
const EVP_CIPHER *EVP_aes_128_gcm(void);

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/modes/ssl_mode.e:
```diff
@@ -0,0 +1,22 @@
+		deferred
+		end
+
+	validate_for_aglorithm (a_algo: SSL_ALGORITHM): BOOLEAN
```
bad Eiffel name for a function returning a boolean.

jvelilla replied:
renamed to `is_valid_for_algorithm`

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/modes/ssl_mode_with_authentication_tag.e:
```diff
@@ -0,0 +1,18 @@
+	tag: detachable MANAGED_POINTER
+			-- The value of the tag (bytes) supplied to the constructor of this mode.
+		note
+			option: stable
```
Is it really needed?
If not, I would remove the `option: stable`

jvelilla replied:
I need to check why I added it

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/modes/ssl_mode_with_initialization_vector.e:
```diff
@@ -0,0 +1,37 @@
+	initialization_vector: detachable MANAGED_POINTER
+			--  The value of the initialization vector for this mode as bytes.
+		note
+			option: stable
```
is option stable really needed? if not, I would remove it.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/modes/ssl_mode_with_initialization_vector.e:
```diff
@@ -0,0 +1,37 @@
+
+feature -- Status Report
+
+	check_iv_length (a_algo: SSL_ALGORITHM): BOOLEAN
```
bad Eiffel name for a function returning a boolean

jvelilla replied:
renamed to `is_valid_iv_length`

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/modes/ssl_mode_with_tweak.e:
```diff
@@ -0,0 +1,13 @@
+note
+	description: "Summary description for {SSL_MODE_WITH_TWEAK}."
```
what is a TWEAK in this context?

jvelilla replied:
it depends on the mode, not all modes needs a TWEAK
Example: https://cryptography.io/en/latest/hazmat/primitives/symmetric-encryption/?highlight=XTS

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/modes/ssl_gcm_mode.e:
```diff
@@ -0,0 +1,71 @@
+
+feature -- Validate algorithm
+
+	validate_for_aglorithm (a_algo: SSL_ALGORITHM): BOOLEAN
```
bad name and typo  aglo vs aglo

jvelilla replied:
I see, what about `is_valid_for_algorithm`?

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/to_remove/ssl_evp.e:
```diff
@@ -0,0 +1,191 @@
+note
```
not reviewed as it is in folder "to_remove"

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/to_remove/ssl_gcm.e:
```diff
@@ -0,0 +1,252 @@
+note
```
not reviewed as it is in folder "to_remove"

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/wrapper/ssl_cipher_context_externals.e:
```diff
@@ -0,0 +1,287 @@
+		                                                   default_pointer,
+		                                                   operation)
+
+		        check success: l_res /= 0 end
```
what is l_res is 0 ?
Should it report an error somewhere?

jvelilla replied:
Yes, still not sure how to handle it.  Any suggestion?

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/wrapper/ssl_cipher_context_externals.e:
```diff
@@ -0,0 +1,287 @@
+
+		     	check success: l_res /= 0 end
+
+				if attached {SSL_GCM_MODE} mode as l_mode then
```
bad indentation

jocelyn replied:
the rest of the routine as well ... hard to review.

jvelilla replied:
fixed

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/wrapper/ssl_cipher_context_externals.e:
```diff
@@ -0,0 +1,287 @@
+
+feature -- Access
+
+	ENCRYPT: INTEGER = 1
```
it seems a constants ... maybe rename the feature, otherwise one may get confuse with eventual "encrypt" routine in other interfaces.

Also move it to a clause:  feature -- Constants

jvelilla replied:
done.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/wrapper/ssl_cipher_context_externals.e:
```diff
@@ -0,0 +1,287 @@
+	mode: SSL_MODE
+			-- current mode to be used.
+
+	operation: INTEGER
```
Specific which value the `operation` can be set to.
And maybe provide a `is_encrypt_operation: BOOLEAN` ....

jvelilla replied:
done.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/wrapper/ssl_cipher_context_externals.e:
```diff
@@ -0,0 +1,287 @@
+	operation: INTEGER
+			-- encrypt or decrypt.
+
+	block_size_bytes: INTEGER
```
better name ?  `block_size`? not sure what the "bytes" has to do here.

jvelilla replied:
maybe, not sure, at the moment I've used the same convention as Python code,
The cipher algorithms represent `block_size` in bits, but the implementation uses the block size in bytes.

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/wrapper/ssl_cipher_context_externals.e:
```diff
@@ -0,0 +1,287 @@
+	iv_nonce: detachable MANAGED_POINTER
+			-- initialization vector.
+
+	finalized : BOOLEAN
```
feature -- Status report

jvelilla replied:
fixed

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/wrapper/ssl_cipher_context_externals.e:
```diff
@@ -0,0 +1,287 @@
+			if
+				attached {SSL_CIPHER_ALGORITHM} a_cipher as l_algo
+			then
+				create l_name.make_from_string (Cipher_name_template)
```
I don't see the benefit to use cipher_name_template`cipher_name_template` here?
it can not be changed or redefined ... and `replace_substring_all` sounds a bit too much.
why not doint

create l_name.make_from_string (l_algo.name)
l_name.extend ('-')
l_name.append (l_algo.key_size.out)
l_name.extend ('-')
l_name.append (a_mode.name)

l_name.to_lower
?

jocelyn commented on Src/unstable/library/network/openssl/src/crypto/wrapper/ssl_error_externals.e:
```diff
@@ -0,0 +1,67 @@
+
+feature -- Constants
+
+	ERR_LIB_EVP: INTEGER = 6
```
maybe explain why 6 ? is it hardcoded value of C macro ?

same for next constant, 138 ?

jvelilla replied:
yes, it's a hardcoded value defined in a C header as macro.
defined as # define ERR_LIB_EVP             6
I've opted to hardcoded the value instead of using a c external call, but maybe I will refactor it and use a C external call, because if the value change we don't need to update the code.

jocelyn commented on Src/unstable/library/network/openssl/src/ssl.e:
```diff
@@ -278,6 +326,22 @@ feature {NONE} -- Externals
+
+
 note
 	copyright: "Copyright (c) 1984-2016, Eiffel Software and others"
```
update copyright year.

jocelyn commented on Src/unstable/library/network/openssl/src/ssl_error_factory.e:
```diff
@@ -0,0 +1,33 @@
+
+feature -- Errors
+
+	retrieve_errors: detachable LIST [SSL_ERROR]
```
bad name for a function

jocelyn commented on Src/unstable/library/network/openssl/test/example_gcm.e:
```diff
@@ -0,0 +1,131 @@
+		do
+
+				-- Create and initialise the context
+			l_ctx := {SSL_EVP}.c_evp_cipher_ctx_new
```
SSL_EVP is in folder "to_remove" ... please also move related tests to a folder "to_remove".
As I guess this is useless to review code that is about to be removed.

jocelyn commented on Src/unstable/library/network/openssl/test/test_aes_mode_gcm.e:
```diff
@@ -0,0 +1,109 @@
+			create l_mode.make ("8b23299fde174053f3d652ba", Void)
+			create cipher.make (l_algo, l_mode)
+			l_encryptor := cipher.encryptor
+			l_computed_ct := l_encryptor.update (create {MANAGED_POINTER}.make_from_array ((create {BYTE_ARRAY_CONVERTER}.make_from_hex_string ("28286a321293253c3e0aa2704a278032")).to_natural_8_array))
```
Note that BYTE_ARRAY_CONVERTER is for now hidden (as implementation).
So user won't be able to use it directly.
then, the interface should have several version of `update` ... `update_with_string`, `update_with_hexadecimal_string` and so on ...

jocelyn replied:
I am not sure if `l_computed_ct` as a MANAGED_POINTER is really user friendly.
Could we have higher abstraction? 
maybe the l_encryptor should keep an attribute with the internal computed result after `update` ...

I think this is what we have in `crypto` lib for HMAC_SHA256 and related.

jocelyn commented on Src/unstable/library/network/socket/netssl/examples/non_blocking/client_server/client/echo_client.rc:
```diff
@@ -0,0 +1,6 @@
+#include <windows.h>
```
do not commit generated .rc files

jocelyn reviewed (commented):
Major issue related to visibility of BYTE_ARRAYED_CONVERTER.
Many functions should be renamed to follow Eiffel rules.

and a few cosmetic points.

jvelilla committed:
Updated copyright.
Clean Code.
Updated SSL_CIPHER_INTERFACE's to follow Eiffel Style.
Refactor rename features to follow Eiffel Style.

jvelilla committed:
Updated directory structure to describe symetric and asymetric encryption
Clean code.
Remove obsolete cluster: to_remove.

jvelilla committed:
Merge branch 'master' of https://github.com/EiffelSoftware/EiffelStudio into es_netssl

jvelilla committed:
Merge branch 'master' of https://github.com/EiffelSoftware/EiffelStudio into es_netssl

jvelilla committed:
Updated OpenSSL library to latest release 1.1.0g
Updated OpenSSL dynamic ecf to use the latest dynamic library.
Updated OpenSSL readme with information about how to compile it (generate static and dynamic library) and how
to install it on Ubuntu (Debian) Linux.
Updated Delivery with the latest OpenSSL libraries for Visual Studio C++ win64 and win32
